src/liballoc/sync.rs

```diff
@@ -328,6 +328,70 @@ impl<T> Arc<T> {
         Self::from_inner(Box::leak(x).into())
     }
 
+    /// Constructs a new `Arc<T>` using a weak reference to itself. Attempting
+    /// to upgrade the weak reference before this function returns will result
+    /// in a `None` value. However, the weak reference may be cloned freely and
+    /// stored for use at a later time.
+    ///
+    /// # Examples
+    /// ```
+    /// #![feature(arc_new_with)]
+    /// #![allow(dead_code)]
+    ///
+    /// use std::sync::{Arc, Weak};
+    ///
+    /// struct Foo {
+    ///     me: Weak<Foo>,
+    /// }
+    ///
+    /// let foo = Arc::new_with(|me| Foo {
+    ///     me: me.clone(),
+    /// });
+    /// ```
+    #[inline]
+    #[unstable(feature = "arc_new_with", issue = "none")]
+    pub fn new_with(data_fn: impl FnOnce(&Weak<T>) -> T) -> Arc<T> {
+        // Construct the inner in the "uninitialized" state with a single
+        // weak reference.
+        let uninit_ptr: NonNull<_> = Box::leak(box ArcInner {
+            strong: atomic::AtomicUsize::new(0),
+            weak: atomic::AtomicUsize::new(1),
+            data: mem::MaybeUninit::<T>::uninit(),
+        })
+        .into();
+        let init_ptr: NonNull<ArcInner<T>> = uninit_ptr.cast();
+
+        let weak = Weak { ptr: init_ptr };
+
+        // It's important we don't give up ownership of the weak pointer, or
+        // else the memory might be freed by the time `data_fn` returns. If
+        // we really wanted to pass ownership, we could create an additional
+        // weak pointer for ourselves, but this would result in additional
+        // updates to the weak reference count which might not be necessary
+        // otherwise.
+        let data = data_fn(&weak);
+
+        // Now we can properly initialize the inner value and turn our weak
+        // reference into a strong reference.
+        unsafe {
+            let inner = init_ptr.as_ptr();
+            ptr::write(&raw mut (*inner).data, data);
+
+            // The above write to the data field must be visible to any threads
+            // which observe a non-zero strong count. This synchronizes with
+            // the `compare_exchange_weak` in `Weak::upgrade`.
```

Comment:
This explains the "release" part. But IMO the more interesting thing to explain is why it is okay to *not* also have an "Acquire" ordering here. That is, the naive thing I would expect is `AcqRel`. (Everything being release-acquire results in a very intuitive ownership-passing-style semantics. Every time a weaker or stronger ordering is used, there's something complicated going on that IMO needs explanation.)

Reply:
Let me try, and if you're happy I'll add it to the comment:

There aren't any side effects we need to "Acquire", because the only capability of a reference to a non-upgradeable `Weak` pointer is to change the "weak count", and that has no effect on the function's behaviour.

Reply:
By "the only capability" here you mean, the only thing `data_fn` could have done, right? If you add that, I agree this makes sense.

Reply:
But this really should be carefully checked by someone else... I know enough about weak memory to be really careful but wouldn't trust myself to actually use a relaxed load here. I'd personally use `AcqRel` just to not have to think about this.

Reply:
> By "the only capability" here you mean, the only thing data_fn could have done, right? If you add that, I agree this makes sense.

Yeah, although it's not really `data_fn` itself: data_fn could clone the Weak and then pass that off to another thread that does something with it. By capability I mean anything that can be done with the `&Weak` or anything derived from that `&Weak`, transitively.

> But this really should be carefully checked by someone else...

Yep.

> I'd personally use AcqRel just to not have to think about this.

IME that's dangerous, particularly in situations where there are multiple atomic variables. If there's a mistake in my reasoning and `Release` semantics are also required, then it's just as likely to me that `SeqCst` is required, or even that no ordering is strong enough and that the code needs to change.

Reply:
Well I've updated the comment in the code as discussed.

> SeqCst is hardly ever required.

You missed off the second part of that sentence 😛 I said SeqCst *or that the code needs to change*. I think the only part I'm disagreeing with you on is that it is generally sufficient to use `AcqRel` everywhere and be confident it's correct: in my experience as soon as more than one atomic variable is involved, you can't get away with that simplification anymore.

With a single atomic variable and AcqRel everywhere, the complexity grows linearly with the amount of code, and things are still fairly intuitive. Once you have more than one atomic variable using AcqRel, the complexity grows exponentially: each atomic variable will have some order on operations, but this may be inconsistent with the order observed by accesses to a different atomic variable.

Each Acquire/Release pair only adds a single ordering constraint between two threads. With multiple threads and multiple atomics, this inevitably leaves the set of possible executions wildly unconstrained. SeqCst is unique in that it enforces a total ordering across *all* atomic accesses, not just the variable being accessed, and so it is capable of placing constraints between multiple threads at once. As a result it can be much easier to reason about, even if a weaker set of memory orderings may be possible, simply because there are exponentially fewer possible executions.

Reply:
> Each Acquire/Release pair only adds a single ordering constraint between two threads. With multiple threads and multiple atomics, this inevitably leaves the set of possible executions wildly unconstrained. SeqCst is unique in that it enforces a total ordering across all atomic accesses, not just the variable being accessed, and so it is capable of placing constraints between multiple threads at once. As a result it can be much easier to reason about, even if a weaker set of memory orderings may be possible, simply because there are exponentially fewer possible executions.

Reasoning based on `SeqCst` requires reasoning by case distinction: did this go first or that? That is the unique guarantee that `SeqCst` provides. This is inherently exponential in nature, as there are exponentially many ways that a bunch of accesses could be ordered.

There is a better way to reason about concurrent code, which is by thinking of everything in terms of ownership transfer. (And in my research I have shown that [ownership transfer is really all you need](https://people.mpi-sws.org/~dreyer/papers/iris-ground-up/paper.pdf).) That avoids the exponential blowup. And for ownership transfer, `AcqRel` is sufficient. Thus, I disagree with your statement that the guarantee `SeqCst` adds actually helps simplify reasoning in general. (Well, you said "*can* be much easier", but it is hard to argue against there being cases where indeed it is simpler. If your statement is that such cases *exist*, I cannot disagree. I am arguing they are not common.) Only rarely does reasoning actually require doing a case distinction which of two operations went first, and only then `SeqCst` should be used.

Anyway, we are digressing. I'd love to have this discussion with you via a better medium than text at some point. :)

Reply:
From correctness perspective it looks fine, although it is unfortunate that everyone using `Weak::upgrade` has to pay for additional synchronization.

> ... once the count hits 0 it must never be above 0.

It would be nice to update / reword this comment in `Weak::upgrade`, since `Arc::new_with` makes it possible to observe strong count going from zero to non-zero.

Reply:
Fixed.